src/plugins/discover/public/application/view_components/canvas/index.tsx

```diff
@@ -20,15 +20,21 @@ import { useOpenSearchDashboards } from '../../../../../opensearch_dashboards_re
 import { filterColumns } from '../utils/filter_columns';
 import { DEFAULT_COLUMNS_SETTING, MODIFY_COLUMNS_ON_SWITCH } from '../../../../common';
 import { OpenSearchSearchHit } from '../../../application/doc_views/doc_views_types';
+import { buildColumns } from '../../utils/columns';
 import './discover_canvas.scss';
 
 // eslint-disable-next-line import/no-default-export
 export default function DiscoverCanvas({ setHeaderActionMenu, history }: ViewProps) {
   const { data$, refetch$, indexPattern } = useDiscoverContext();
-  const {
-    services: { uiSettings },
-  } = useOpenSearchDashboards<DiscoverViewServices>();
-  const { columns } = useSelector((state) => state.discover);
+  const { services: {uiSettings, capabilities} } = useOpenSearchDashboards<DiscoverViewServices>();
+  const { columns } = useSelector(state => {
+    const stateColumns = state.discover.columns;
+  
+    // check if stateColumns is not undefined, otherwise use buildColumns
+    return {
+      columns: stateColumns !== undefined ? stateColumns : buildColumns([])
+    };
+  });
```

Comment:
Same goes here. Initializing them in the redux store means that you can always expect these to be not undefined and avoid this check alltogether.